src/main/scala/mesosphere/marathon/core/storage/migration/Migration.scala

```diff
@@ -0,0 +1,208 @@
+import mesosphere.marathon.core.storage.store.PersistenceStore
+import mesosphere.marathon.metrics.Metrics
+import mesosphere.marathon.{ BuildInfo, MigrationFailedException }
+
+import scala.async.Async.{ async, await }
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, Future }
+import scala.util.control.NonFatal
+// scalastyle:on
+
+class Migration(
+    private[migration] val legacyConfig: Option[LegacyStorageConfig],
+    private[migration] val persistenceStore: Option[PersistenceStore[_, _, _]],
+    private[migration] val appRepository: AppRepository,
+    private[migration] val groupRepository: GroupRepository,
+    private[migration] val deploymentRepository: DeploymentRepository,
+    private[migration] val taskRepo: TaskRepository,
+    private[migration] val taskFailureRepo: TaskFailureRepository)(implicit
+  mat: Materializer,
+    metrics: Metrics) extends StrictLogging {
+  //scalastyle:off magic.number
+
+  import StorageVersions._
+  import Migration._
+
+  type MigrationAction = (StorageVersion, () => Future[Any])
+
+  private[migration] val minSupportedStorageVersion = StorageVersions(0, 8, 0)
+
+  private[migration] val legacyStoreFuture: Future[Option[PersistentStore]] = legacyConfig.map { config =>
+    val store = config.store
+    store match {
+      case s: PersistentStoreManagement =>
+        s.initialize().map(_ => Some(store))
+      case _ =>
+        Future.successful(Some(store))
+    }
+  }.getOrElse(Future.successful(None))
+
+  /**
+    * All the migrations, that have to be applied.
+    * They get applied after the master has been elected.
+    */
+  def migrations: List[MigrationAction] =
+    List(
+      StorageVersions(0, 7, 0) -> { () =>
+        Future.failed(new IllegalStateException("migration from 0.7.x not supported anymore"))
+      },
+      StorageVersions(0, 11, 0) -> { () =>
+        new MigrationTo0_11(legacyConfig).migrateApps().recover {
+          case NonFatal(e) => throw new MigrationFailedException("while migrating storage to 0.11", e)
+        }
+      },
+      StorageVersions(0, 13, 0) -> { () =>
+        new MigrationTo0_13(legacyConfig).migrate().recover {
+          case NonFatal(e) => throw new MigrationFailedException("while migrating storage to 0.13", e)
+        }
+      },
+      StorageVersions(0, 16, 0) -> { () =>
+        new MigrationTo0_16(legacyConfig).migrate().recover {
+          case NonFatal(e) => throw new MigrationFailedException("while migrating storage to 0.16", e)
+        }
+      },
+      StorageVersions(1, 2, 0) -> { () =>
+        new MigrationTo1_2(legacyConfig).migrate().recover {
+          case NonFatal(e) => throw new MigrationFailedException("while migrating storage to 1.2", e)
+        }
+      },
+      StorageVersions(1, 3, 0) -> { () =>
```

Comment:
1) If you want cherry pick this in 1.2, we need to change this to the correct version. 
2) since we now have 2 different layouts in one version (legacy and new), it is not clear from the version, which layout has been used. Perhaps we also have to store the layout in the SchemaVersion. (Otherwise it is hard to say what to do, once we have 1.3)